syncer/db.go

```diff
@@ -196,57 +253,35 @@ func (conn *Conn) executeSQLWithIgnore(tctx *tcontext.Context, ignoreError func(
 	return ret.(int), nil
 }
 
-func (conn *Conn) executeSQL(tctx *tcontext.Context, queries []string, args ...[]interface{}) (int, error) {
+func (conn *DBConn) executeSQL(tctx *tcontext.Context, queries []string, args ...[]interface{}) (int, error) {
 	return conn.executeSQLWithIgnore(tctx, nil, queries, args...)
 }
 
-func createBaseConn(dbCfg config.DBConfig, timeout string, rawDBCfg *baseconn.RawDBConfig) (*baseconn.BaseConn, error) {
-	dbDSN := fmt.Sprintf("%s:%s@tcp(%s:%d)/?charset=utf8mb4&interpolateParams=true&readTimeout=%s&maxAllowedPacket=%d",
-		dbCfg.User, dbCfg.Password, dbCfg.Host, dbCfg.Port, timeout, *dbCfg.MaxAllowedPacket)
-	baseConn, err := baseconn.NewBaseConn(dbDSN, &retry.FiniteRetryStrategy{}, rawDBCfg)
-	if err != nil {
-		return nil, terror.DBErrorAdapt(err, terror.ErrDBDriverError)
-	}
-	return baseConn, nil
-}
-
-func createConn(cfg *config.SubTaskConfig, dbCfg config.DBConfig, timeout string) (*Conn, error) {
-	baseConn, err := createBaseConn(dbCfg, timeout, baseconn.DefaultRawDBConfig())
-	if err != nil {
-		return nil, err
-	}
-	return &Conn{baseConn: baseConn, cfg: cfg}, nil
-}
-
-func createConns(cfg *config.SubTaskConfig, dbCfg config.DBConfig, count int, timeout string) ([]*Conn, error) {
-	dbs := make([]*Conn, 0, count)
-
-	rawDBCfg := &baseconn.RawDBConfig{
-		MaxIdleConns: cfg.SyncerConfig.WorkerCount,
-	}
-	baseConn, err := createBaseConn(dbCfg, timeout, rawDBCfg)
+func createConns(tctx *tcontext.Context, cfg *config.SubTaskConfig, dbCfg config.DBConfig, count int) (*conn.BaseDB, []*DBConn, error) {
+	conns := make([]*DBConn, 0, count)
+	db, err := createBaseDB(dbCfg)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 	for i := 0; i < count; i++ {
-		// TODO use *sql.Conn instead of *sql.DB
-		// share db by all conns
-		bc := &baseconn.BaseConn{baseConn.DB, baseConn.DSN, baseConn.RetryStrategy, rawDBCfg}
-		dbs = append(dbs, &Conn{baseConn: bc, cfg: cfg})
-	}
-	return dbs, nil
-}
-
-func closeConns(tctx *tcontext.Context, conns ...*Conn) {
-	for _, conn := range conns {
-		err := conn.baseConn.Close()
+		dbConn, err := db.GetBaseConn(tctx.Context())
 		if err != nil {
-			tctx.L().Error("fail to close baseConn connection", log.ShortError(err))
+			closeBaseDB(tctx, db)
+			return nil, nil, terror.WithScope(err, terror.ScopeDownstream)
+		}
+		resetConnFn := func(tctx *tcontext.Context) (*conn.BaseConn, error) {
+			err := db.CloseBaseConn(dbConn)
```

Comment:
after some reset operations triggered, is this `dbConn` refer to the newest one (re-created after reset) or the oldest one (created in L267 for the first time)?

Reply:
yes，you are right，after I tested，dbConn refer to newest one，I change the reset function here，please check [loader](https://github.com/pingcap/dm/pull/266/files#diff-5f5bb90e4d87dad0cb25c377bb10679bR194) and [syncer](https://github.com/pingcap/dm/pull/266/files#diff-c3cc6fccd12f8e1679a575d1aaf8fd14R272)